### PR TITLE
add: cleanup .gitignore decorator

### LINKED ISSUE #1641: dvc add is not atomic command
If dvc add fails it keeps new created `.gitignore`. See repro script from #1640.

DVC have to revert all the changes after failing command or output warning if it is not possible.

### PR DESCRIPTION
**Pull request by pared**
Fixes #1641

### CODE REVIEW

File: dvc/repo/add.py [hunk 1]
```diff
@@ -11,6 +27,8 @@ def add(repo, fname, recursive=False, no_commit=False):
 
     for stage in stages:
         stage.dump()
+    for stage in stages:
+        repo.scm.ignore_remove(stage.path)
```
Comment (efiop) [inline dvc/repo/add.py]: Why are we removing stage.path from gitignore at all?

File: dvc/repo/add.py [hunk 2]
```diff
@@ -2,6 +2,22 @@
 from dvc.stage import Stage
 
 
+def cleanup_gitignore(add_method):
```
Comment (efiop) [inline dvc/repo/add.py]: Let's move this to `dvc/scm`. Also, since we do have mercurial support upcoming, maybe we could name it something not git-specific?

File: dvc/repo/add.py [hunk 3]
```diff
@@ -2,6 +2,22 @@
...
+    def run(*args, **kw):
+        repo = args[0]
```
Comment (efiop) [inline dvc/repo/add.py]: Also, probably worth adding an assert that would make sure that repo is actually a Repo() instance.

File: dvc/repo/add.py (hunk 1 above)
Comment (pared) [reply dvc/repo/add.py]: Sorry, leftover after initial changes

File: dvc/scm/git.py [hunk 4]
```diff
@@ -71,6 +73,7 @@ def _get_gitignore(self, path):
         return entry, gitignore
 
     def ignore(self, path):
+        self.ignored_paths.append(path)
```
Comment (efiop) [inline dvc/scm/git.py]: Let's add it to the list _after_ we've added it to gitignore itself. E.g. it might be that we already have an entry for it, so we shouldn't remove what we didn't add. Looks like it is worth adding a test for it.

File: dvc/repo/run.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 
 def run(
-    self,
+    repo,
```
Comment (efiop) [inline dvc/repo/run.py]: You probably meant to use the same `@revert_ignores_on_failure` decorator for `run()` and `imp()` methods? Or just unresistable cosmetic changes? 🙂 It is worth introducing such either in a separate patch or in a separate PR, so that it doesn't obstruct the view. No need to bother splitting it now, but just worth to keep in mind 🙂

File: dvc/scm/base.py [hunk 6]
```diff
@@ -139,3 +139,15 @@ def list_tags(self):  # pylint: disable=no-self-use
 
     def install(self):
         """Adds dvc commands to SCM hooks for the repo."""
+
+    def cleanup_ignores(self):
+        """
+        This method should clean up ignores (eg. entries in .gitignore),
+        use, when method editing ignores (eg. add, run, import) fails to
+        perform its task
```
Comment (efiop) [inline dvc/scm/base.py]:
```suggestion
        perform its task.
```

File: dvc/repo/run.py (hunk 5 above)
Comment (pared) [reply dvc/repo/run.py]: You are right, I wanted to introduce annotation here. Reverting

File: dvc/scm/__init__.py [hunk 7]
```diff
@@ -21,3 +23,20 @@ def SCM(root_dir, repo=None):  # pylint: disable=invalid-name
...
+
+def scm_context(method):
+    def run(*args, **kw):
+        repo = args[0]
+        assert type(repo) is dvc.repo.Repo
+        scm = repo.scm
+        try:
+            result = method(*args, **kw)
+            scm.reset_ignores()
+            repo.remind_to_git_add()
```
Comment (efiop) [inline dvc/scm/__init__.py]: Let's move `remind_to_git_add()` along with `repo.files_to_git_add` into scm instead of keeping it in the repo. That should fit nicely 🙂

File: dvc/repo/add.py [hunk 8]
```diff
@@ -1,7 +1,10 @@
 import os
+
+from dvc.scm import scm_context
 from dvc.stage import Stage
 
 
+@scm_context
```
Comment (efiop) [inline dvc/repo/add.py]: Shouldn't we use it in `repo.run()` and `repo.imp()` as well?

Comment (pared) [reply dvc/repo/add.py]: I guess we could, however, corrupting stage as in test for add does not result in dangling edited .gitignore. Do you think we should also have tests for `run` and `imp`?

File: dvc/repo/imp.py [hunk 9]
```diff
@@ -8,12 +8,11 @@ def imp(self, url, out, resume=False):
 
     self.check_dag(self.stages() + [stage])
 
-    self.files_to_git_add = []
```
Comment (efiop) [inline dvc/repo/imp.py]: But you are not using the decorator for imp().

File: dvc/repo/move.py [hunk 10]
```diff
@@ -65,4 +65,4 @@ def move(self, from_path, to_path):
 
     stage.dump()
 
-    self.remind_to_git_add()
```
Comment (efiop) [inline dvc/repo/move.py]: Also not using scm_context for move(), so user won't be reminded.

File: dvc/repo/reproduce.py [hunk 11]
```diff
@@ -65,8 +65,6 @@ def reproduce(
     else:
         targets.append(target)
 
-    self.files_to_git_add = []
```
Comment (efiop) [inline dvc/repo/reproduce.py]: Same here.

File: dvc/repo/run.py [hunk 12]
```diff
@@ -53,13 +53,12 @@ def run(
 
     self.check_dag(self.stages() + [stage])
 
-    self.files_to_git_add = []
```
Comment (efiop) [inline dvc/repo/run.py]: And here.

File: dvc/repo/imp.py (hunk 9 above)
Comment (pared) [reply dvc/repo/imp.py]: That is true, however take a look at stage.dump(), there is code for file tracking. Even without scm_context, files_to_git_add is unnecessary here.

File: dvc/repo/move.py (hunk 10 above)
Comment (pared) [reply dvc/repo/move.py]: As above.

File: dvc/repo/reproduce.py (hunk 11 above)
Comment (pared) [reply dvc/repo/reproduce.py]: As above.

File: dvc/repo/imp.py (hunk 9 above)
Comment (pared) [reply dvc/repo/imp.py]: As discussed, wrapping with scm_context.

File: dvc/repo/move.py (hunk 10 above)
Comment (pared) [reply dvc/repo/move.py]: As discussed, wrapping with scm_context.

File: dvc/repo/reproduce.py (hunk 11 above)
Comment (pared) [reply dvc/repo/reproduce.py]: As discussed, wrapping with scm_context.

File: dvc/repo/run.py (hunk 12 above)
Comment (pared) [reply dvc/repo/run.py]: As discussed, wrapping with scm_context.
